keps/sig-instrumentation/draft-20190404-kubernetes-control-plane-metrics-stability.md

```diff
@@ -0,0 +1,252 @@
+
+```go
+var deprecatedMetric = kubemetrics.NewCounterVec( // this is a wrapped initializer, which takes in our custom metric definitions
+    deprecatedMetricDefinition, // this is our custom wrapped metric definition from above
+    []string{"some-label", "other-label"},
+}
+var alphaMetric = kubemetrics.NewCounterVec{
+    alphaMetricDefinition, // this is also our custom wrapped metric definition from above
+    []string{"some-label", "other-label"},
+}
+```
+
+### Metric Enrollment Phase
+
+Currently, metric enrollment involves calls to a prometheus function which enrolls the metric in a global registry, like so:
+```go
+prometheus.MustRegister(someCounterVecMetric)
+```
+
+Wrapping a prometheus registry with a kubernetes specific one, would allow us to take our custom metrics from our instantiation phase and execute custom logic based on our custom metadata. Our custom registry would hold a reference to a prometheus registry and defer metric enrollment unless preconditions were met:
+
+```go
+import version "k8s.io/apimachinery/pkg/version"
+
+type Registry struct {
+    promregistry *prometheus.Registry
+    KubeVersion version.Info
+}
+
+// inject custom registration behavior into our registry wrapper
+func (r *Registry) MustRegister(metric kubemetrics.Metric) {
+    // pretend we have a version comparison utility library
+    if metricutils.compare(metric.DeprecatedVersion).isLessThan(r.KubeVersion) {
+        // check if binary has deprecated metrics enabled otherwise
+        // no-op registration
+        return
+    } else if metricutils.compare(metric.DeprecatedVersion).isEqual(r.KubeVersion) {
+        // append deprecated text to description
+        // emit warning in logs
+        // continue to actual registration
+    }
+    // append alpha text to metric description if metric.isAlpha
+    // fallback to original prometheus behavior
+    r.promregistry.MustRegister(metric.realMetric)
+}
+
+```
+
+Which we would invoke, like so:
+```go
+kubemetrics.MustRegister(deprecatedMetric)
+kubemetrics.MustRegister(alphaMetric)
+```
+
+## Stability Classes
+
+This proposal introduces three stability classes for metrics: (1) Alpha, (2) Stable, (3) Deprecated. These classes are intended to make explicit the API contract between the control-plane and the ingester of control-plane metrics.
+
+__Alpha__ metrics have __*no*__ stability guarantees; as such they can be modified or deleted at any time. At this time, all kubernetes metrics implicitly fall into this category.
+
+__Stable__ metrics can be guaranteed to *not change*, except that the metric may become marked deprecated for a future kubernetes version. By *not change*, we mean three things:
+
+1. the metric itself will not be deleted
+3. the type of metric will not be modified
+4. no labels can be added or removed from this metric
```

Comment:
maybe worth linking to the discussion or expounding a bit on this.

FWIW, headings get turned into anchors, so I usually use links + anchors instead of footnotes:

```markdown
[discussed below](#expound-on-the-thing)

### Expound on the Thing

More details were necessary to make people understand why versioning is useful.
```

Reply:
Ah, perfect! I'll use that.